x/gamm/twap/store.go

```diff
@@ -51,8 +51,18 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
 	osmoutils.MustSet(store, key2, &twap)
 }
 
-func (k Keeper) pruneRecordsBeforeTime(ctx sdk.Context, lastTime time.Time) {
-	// TODO: Stub
+func (k Keeper) pruneRecordsBeforeTime(ctx sdk.Context, lastTime time.Time) error {
+	store := ctx.KVStore(k.storeKey)
+	iter := store.Iterator([]byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastTime, 0, "", ""))
+	defer iter.Close()
+	for ; iter.Valid(); iter.Next() {
+		twapToRemove, err := types.ParseTwapFromBz(iter.Value())
+		if err != nil {
+			return err
+		}
+		k.deleteHistoricalRecord(ctx, twapToRemove)
+	}
```

Comment:
Can we push this to a helper in osmo_utils/store ?

Also I'm not entirely sure if your allowed to delete keys while iterating over them or if you have to queue all the keys and delete in a second pass over it. I mostly have just forgotten / mixed up a lot of the oddiites of the cache kv store

Reply:
> not entirely sure if your allowed to delete keys while iterating over them or if you have to queue all the keys and delete in a second pass over it

I looked into the underlying implementation. So, what seems to be happening on `Delete()` is that we:
- mark a record as deleted in cache: https://github.com/osmosis-labs/cosmos-sdk/blob/f09814c9e23fd5832b11666a871bf51f36f85e41/store/cachekv/store.go#L99
- then, when calling `iter.Next()`, it skips until a record exists or iterator becomes invalid:https://github.com/osmosis-labs/cosmos-sdk/blob/f09814c9e23fd5832b11666a871bf51f36f85e41/store/cachekv/mergeiterator.go#L63

We do not write into IAVL until commit so the deletions are handled in cache.
Additionally, all tests passing makes me more confident that there is no problem with deleting while iterating.

Reply:
> Can we push this to a helper in osmo_utils/store ?

I'm not sure if I can generalize this well enough to justify the move to `osmo_utils/store`. 

I tried this:
```
func iterate(store sdk.KVStore, from []byte, to []byte, fn func() error) error {
	iter := store.Iterator(from, to)
	defer iter.Close()
	for ; iter.Valid(); iter.Next() {
		if err := fn(); err != nil {
			return err
		}
	}
	return nil
}
```

So that the method in question would look like:
```
func (k Keeper) pruneRecordsBeforeTime(ctx sdk.Context, lastTime time.Time) error {
	store := ctx.KVStore(k.storeKey)

	err := iterate(store, []byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastTime, 0, "", ""), func() error {
		return nil
	})

	if err != nil {
		return err
	}
```

However, I think this abstraction only adds unnecessary complexity without reducing that much boilerplate. Please let me know what you think

Reply:
Ah agreed your right - I was thinking that it was deleting all keys in the range. Didn't realize it was actually calling another methoid for delete, that deleted two things.